src/Nethermind/Nethermind.TxPool/FilteredTxPool.cs

```diff
@@ -0,0 +1,58 @@
+﻿//  Copyright (c) 2018 Demerzel Solutions Limited
+//  This file is part of the Nethermind library.
+// 
+//  The Nethermind library is free software: you can redistribute it and/or modify
+//  it under the terms of the GNU Lesser General Public License as published by
+//  the Free Software Foundation, either version 3 of the License, or
+//  (at your option) any later version.
+// 
+//  The Nethermind library is distributed in the hope that it will be useful,
+//  but WITHOUT ANY WARRANTY; without even the implied warranty of
+//  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+//  GNU Lesser General Public License for more details.
+// 
+//  You should have received a copy of the GNU Lesser General Public License
+//  along with the Nethermind. If not, see <http://www.gnu.org/licenses/>.
+// 
+
+using System.Collections.Generic;
+using Nethermind.Core;
+using Nethermind.Core.Specs;
+using Nethermind.Crypto;
+using Nethermind.Logging;
+using Nethermind.State;
+
+namespace Nethermind.TxPool
+{
+    public class FilteredTxPool : TxPool
+    {
+        private readonly ITxPoolFilter _txPoolFilter;
+
+        public interface ITxPoolFilter
+        {
+            bool Accept(Transaction tx);
+        }
+        
+        public FilteredTxPool(
+            ITxStorage txStorage, 
```

Comment:
can we have other checks implemented as filters now?
For example:
 NonceFilter(StateProvider)
 ChainidFilter()
 NullSignatureFilter()
----
or we can create a pipeline ->
  RecoverSignature, Filter, etc...

This would make TxPools more composable and easier to read
Now filtering responsibility is split between filters and the pool itself

Reply:
Yes we can, but I wouldn't make filters that are integral to TxPool functionality composable and optional, TxPool should still force its basic rules, in the end it won't give us much benefit, but the code might be just a bit nicer.

Reply:
They are actually not integral. BeamSync pool cannot verify everything properly (nonces and so on)